src/bin/ion/commands/beta/generate/utils.rs

```diff
@@ -0,0 +1,158 @@
+        match self {
+            Language::Rust => "rs",
+            Language::Java => "java",
+        }
+    }
+
+    pub fn file_name(&self, name: &str) -> String {
+        match self {
+            Language::Rust => name.to_case(Case::Snake),
+            Language::Java => name.to_case(Case::UpperCamel),
+        }
+    }
+}
+
+impl From<&str> for Language {
+    fn from(value: &str) -> Self {
+        match value {
+            "java" => Language::Java,
+            "rust" => Language::Rust,
+            _ => unreachable!("Unsupported programming language: {}, this tool only supports Java and Rust code generation.", value)
+        }
+    }
+}
+
+impl Display for Language {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Language::Rust => {
+                write!(f, "rust")
+            }
+            Language::Java => {
+                write!(f, "java")
+            }
+        }
+    }
+}
+
+/// Represents a context that will be used for code generation
+pub struct CodeGenContext {
+    // Represents the templating engine - tera
+    pub(crate) tera: Tera,
+    pub(crate) language: Language,
+    // Initially the data_model field is set to None.
+    // Once an ISL type definition is mapped to a data model this will have Some value.
+    pub(crate) data_model: Option<DataModel>,
+    // Represents a counter for naming anonymous type definitions
+    pub(crate) anonymous_type_counter: usize,
+}
+
+impl CodeGenContext {
+    pub fn new(language: Language) -> Self {
+        Self {
+            language,
+            data_model: None,
+            anonymous_type_counter: 0,
+            tera: Tera::new("src/bin/ion/commands/beta/generate/templates/**/*.templ").unwrap(),
+        }
+    }
+
+    pub fn with_data_model(&mut self, data_model: DataModel) {
+        self.data_model = Some(data_model);
+    }
+
+    pub fn with_initial_data_model(&mut self) {
+        // Initially the data model is set to None, this will be set with Some(_) value when data model is determined in code generation process
+        self.data_model = None;
+    }
+
+    /// Returns a string that represent the template name based on data model type.
+    pub fn template_name(&self) -> &str {
+        if let Some(data_model) = &self.data_model {
+            return match (&self.language, data_model) {
+                (
+                    Language::Rust,
+                    DataModel::Struct | DataModel::UnitStruct | DataModel::SequenceStruct,
+                ) => "struct",
+                (
+                    Language::Java,
+                    DataModel::Struct | DataModel::UnitStruct | DataModel::SequenceStruct,
+                ) => "class",
+            };
+        }
+        "" // Default value is an empty string
+    }
+}
+
+/// Represents a data model type that can be used to determine which templates can be used for code generation.
+#[derive(Debug, Clone, PartialEq, Serialize)]
+pub enum DataModel {
```

Comment:
🗺️ Currently only includes data models related to `element`, `fields` and `type` constraints. More data models will be added with other constraints support like `enum`.